File: frontend/coordinator-android/app/src/main/java/ug/qaat/coordinator/ui/LoginScreen.kt
package ug.qaat.coordinator.ui

import android.widget.Toast
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import ug.qaat.coordinator.R
import ug.qaat.coordinator.di.Graph
import ug.qaat.coordinator.net.AuthClient
import ug.qaat.coordinator.net.ManifestClient
import ug.qaat.coordinator.net.Net
import ug.qaat.coordinator.store.SessionStore

/** Unified KIU QAAT sign-in for every role (coordinator / student / lecturer). Identifier =
 *  email, registration number, or staff ID; the app routes by the returned role. */
@Composable
fun LoginScreen(onLoggedIn: () -> Unit) {
    val scope = rememberCoroutineScope()
    val ctx = LocalContext.current
    var identifier by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var totp by remember { mutableStateOf("") }
    var needsMfa by remember { mutableStateOf(false) }
    var busy by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    // Wake the free-tier backend the moment this screen shows, so its cold start (up to ~1 min)
    // overlaps with the user typing — turning a stalled sign-in into an instant one.
    LaunchedEffect(Unit) { Net.warmUp() }

    Box(Modifier.fillMaxSize()) {
    // Faint institution logo watermark behind everything.
    Image(
        painter = painterResource(R.drawable.qaat_logo),
        contentDescription = null,
        contentScale = ContentScale.Fit,
        modifier = Modifier.align(Alignment.Center).size(340.dp).alpha(0.05f),
    )
    Column(
        Modifier.fillMaxSize().padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Image(
            painter = painterResource(R.drawable.qaat_logo),
            contentDescription = "KIU QAAT",
            modifier = Modifier.size(112.dp),
        )
        Spacer(Modifier.height(14.dp))
        Text("KIU QAAT", style = MaterialTheme.typography.headlineMedium)
        Text("Sign in — staff email, registration number, or staff ID",
            style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 2.dp))
        Spacer(Modifier.height(24.dp))
        OutlinedTextField(identifier, { identifier = it.trim() }, label = { Text("Email / Reg. no / Staff ID") },
            singleLine = true, modifier = Modifier.fillMaxWidth())
        Spacer(Modifier.height(8.dp))
        PasswordField(password, { password = it }, "Password", modifier = Modifier.fillMaxWidth())
        if (needsMfa) {
            Spacer(Modifier.height(8.dp))
            OutlinedTextField(totp, { totp = it }, label = { Text("Authenticator code") }, singleLine = true, modifier = Modifier.fillMaxWidth())
        }
        error?.let { msg ->
            Spacer(Modifier.height(8.dp))
            Text(msg, color = MaterialTheme.colorScheme.error)
            // A "Secure connection failed" almost always means the phone's clock is wrong (phones with
            // no SIM never network-sync their time). Surface the exact fix inline.
            if (msg.contains("date", true) || msg.contains("time", true) || msg.contains("secure connection", true)) {
                Spacer(Modifier.height(6.dp))
                Surface(color = MaterialTheme.colorScheme.errorContainer, shape = MaterialTheme.shapes.small) {
                    Text("Fix: Settings → System → Date & time → turn ON “Set time automatically” (or set today’s correct date), then tap Sign in again. A wrong clock blocks the secure connection.",
                        Modifier.padding(10.dp), style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onErrorContainer)
                }
            }
        }

        Spacer(Modifier.height(16.dp))
        Button(
            enabled = !busy && identifier.isNotBlank() && password.isNotBlank(),
            modifier = Modifier.fillMaxWidth(),
            onClick = {
                busy = true; error = null
                scope.launch {
                    runCatching {
                        val res = AuthClient().appLogin(identifier.trim(), password, totp.ifBlank { null }) { needsMfa = true }
                            ?: return@runCatching   // MFA prompt shown
                        AppState.token = res.token; AppState.userId = res.userId
                        AppState.tenantId = res.tenantId; AppState.deviceBindingKey = res.deviceBindingKey
                        AppState.coordinatorName = res.fullName; AppState.coordinatorEmail = identifier.trim(); AppState.role = res.role
                        AppState.coordinatorTitle = res.title; AppState.coordinatorRegNo = res.registrationNo
                        AppState.studentId = res.studentId.ifBlank { null }
                        AppState.staffId = res.staffId.ifBlank { null }
                        AppState.forcePasswordChange = res.forcePasswordChange
                        // Persist the session IMMEDIATELY so auto-login survives a close.
                        SessionStore.saveSession(res.token, res.userId, res.tenantId, res.deviceBindingKey,
                            res.fullName, identifier.trim(), res.role, res.title, res.registrationNo,
                            res.studentId, res.staffId, "")
                        SessionStore.saveAppCredentials(identifier.trim(), password, "")
                        val greet = res.fullName.ifBlank { "back" }
                        Toast.makeText(ctx, "Welcome, $greet 👋", Toast.LENGTH_LONG).show()
                        onLoggedIn()
                        // Institution branding applies to every role.
                        runCatching {
                            ug.qaat.coordinator.net.BrandingClient(res.token).fetch()?.let {
                                AppState.branding = it; SessionStore.saveBranding(it)
                            }
                        }
                        // Only the COORDINATOR needs the daily manifest (roster + keys for the hub).
                        if (res.role == "COORDINATOR") runCatching {
                            val m = ManifestClient(Graph.db.dao()).fetchAndStore(res.token)
                            AppState.manifest = m; SessionStore.saveManifest(m)
                        }
                    }.onFailure { error = ug.qaat.coordinator.net.Net.friendly(it) }
                    busy = false
                }
            },
        ) { Text(if (busy) "Signing in…" else "Sign in") }
        if (busy) {
            Spacer(Modifier.height(10.dp))
            LinearProgressIndicator(Modifier.fillMaxWidth())
            Spacer(Modifier.height(6.dp))
            Text(
                "Contacting the server… the first sign-in can take up to a minute while it wakes up. Please keep waiting.",
                style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }

        // NOTE: there is deliberately NO editable "server address" field here. Every installed
        // phone always talks to the built-in KIU cloud backend (BuildConfig.API_BASE) so the app
        // works identically on every device with zero setup. A wrong address typed/saved on one
        // phone used to break sign-in with confusing, device-specific errors — that is gone now.
        // (To point a build at a different backend, rebuild with -Pqaat.apiBase=… — it's a
        // build-time choice, never something a user has to configure on the phone.)
    }
    // Footer statement — same as the web apps.
    Text(
        "Powered by LIGHT TECHNOLOGIES",
        style = MaterialTheme.typography.labelSmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.align(Alignment.BottomCenter).padding(bottom = 16.dp),
    )
    }
}

File: frontend/coordinator-flutter/lib/ui/login_screen.dart
import 'package:flutter/material.dart';

import '../net/auth_client.dart';
import '../net/branding.dart';
import '../net/net.dart';
import '../net/session_store.dart';
import 'brand_widgets.dart';

/// Unified KIU QAAT sign-in for every role (coordinator / student / lecturer).
/// Identifier = email, registration number, or staff ID; the app routes by the
/// returned role. A faithful mirror of the native `LoginScreen.kt`: the bundled
/// logo, institution watermark, show/hide password, the clock-fix hint and the
/// LIGHT TECHNOLOGIES footer.
///
/// Two-phase when the account requires TOTP: the first submit answers
/// `MFA_REQUIRED` and reveals the authenticator-code field; the second submit
/// sends `totp_code`.
class LoginScreen extends StatefulWidget {
  const LoginScreen({super.key, this.auth = const AuthClient()});

  final AuthClient auth;

  @override
  State<LoginScreen> createState() => _LoginScreenState();
}

class _LoginScreenState extends State<LoginScreen> {
  final _identifier = TextEditingController();
  final _password = TextEditingController();
  final _totp = TextEditingController();
  bool _needsMfa = false;
  bool _passwordVisible = false;
  bool _busy = false;
  Object? _error;

  @override
  void initState() {
    super.initState();
    // Wake the free-tier backend the moment this screen shows, so its cold start
    // overlaps with the person typing, instead of stalling the sign-in.
    Net.warmUp();
  }

  @override
  void dispose() {
    _identifier.dispose();
    _password.dispose();
    _totp.dispose();
    super.dispose();
  }

  bool get _canSubmit =>
      !_busy && _identifier.text.trim().isNotEmpty && _password.text.isNotEmpty;

  Future<void> _submit() async {
    final identifier = _identifier.text.trim();
    setState(() {
      _busy = true;
      _error = null;
    });
    try {
      final result = await widget.auth.appLogin(
        identifier,
        _password.text,
        totp: _needsMfa ? _totp.text : null,
      );
      if (!mounted) return;
      if (result == null) {
        // MFA_REQUIRED — keep the typed credentials, just ask for the code.
        setState(() => _needsMfa = true);
      } else {
        // Adopt the institution's identity now that we carry a token; failures keep
        // the bundled default. Best-effort, fire-and-forget.
        BrandingState.refresh(result.token);
        // Persist the session so a later launch boots straight into the hub without
        // re-typing credentials (offline-tolerant sign-in).
        await SessionStore.save(result);
        if (!mounted) return;
        Navigator.of(context).pushNamedAndRemoveUntil(
          '/home',
          (_) => false,
          arguments: result,
        );
      }
    } catch (e) {
      if (!mounted) return;
      setState(() => _error = e);
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final branding = BrandingState.current.value;
    final friendly = Net.friendly;
    final errorText = _error == null ? null : friendly(_error!);
    return Scaffold(
      body: Stack(
        children: [
          // Faint institution-logo watermark behind everything.
          const Center(
            child: Opacity(
              opacity: 0.05,
              child: Image(
                image: AssetImage('assets/branding/qaat_logo.png'),
                width: 340,
                height: 340,
                fit: BoxFit.contain,
              ),
            ),
          ),
          Center(
            child: SingleChildScrollView(
              padding: const EdgeInsets.all(24),
              child: ConstrainedBox(
                constraints: const BoxConstraints(maxWidth: 420),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: [
                    const Center(
                      child: BrandLogo(branding: null, size: 112),
                    ),
                    const SizedBox(height: 14),
                    Text('KIU QAAT',
                        textAlign: TextAlign.center,
                        style: theme.textTheme.headlineMedium),
                    const SizedBox(height: 2),
                    Text(
                      'Sign in — staff email, registration number, or staff ID',
                      textAlign: TextAlign.center,
                      style: theme.textTheme.bodySmall
                          ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
                    ),
                    const SizedBox(height: 24),
                    TextField(
                      controller: _identifier,
                      enabled: !_busy,
                      onChanged: (_) => setState(() {}),
                      decoration: const InputDecoration(
                        labelText: 'Email / Reg. no / Staff ID',
                        border: OutlineInputBorder(),
                      ),
                    ),
                    const SizedBox(height: 8),
                    TextField(
                      controller: _password,
                      enabled: !_busy,
                      obscureText: !_passwordVisible,
                      onChanged: (_) => setState(() {}),
                      onSubmitted: (_) => _canSubmit ? _submit() : null,
                      decoration: InputDecoration(
                        labelText: 'Password',
                        border: const OutlineInputBorder(),
                        suffixIcon: TextButton(
                          onPressed: () => setState(
                              () => _passwordVisible = !_passwordVisible),
                          child:
                              Text(_passwordVisible ? 'Hide' : 'Show'),
                        ),
                      ),
                    ),
                    if (_needsMfa) ...[
                      const SizedBox(height: 8),
                      TextField(
                        controller: _totp,
                        enabled: !_busy,
                        keyboardType: TextInputType.number,
                        onChanged: (_) => setState(() {}),
                        onSubmitted: (_) => _canSubmit ? _submit() : null,
                        decoration: const InputDecoration(
                          labelText: 'Authenticator code',
                          border: OutlineInputBorder(),
                        ),
                      ),
                    ],
                    if (errorText != null) ...[
                      const SizedBox(height: 8),
                      Text(errorText,
                          style: TextStyle(color: theme.colorScheme.error)),
                      _ClockFixHint(errorText: errorText),
                    ],
                    const SizedBox(height: 16),
                    FilledButton(
                      onPressed: _canSubmit ? _submit : null,
                      child: Text(_busy ? 'Signing in…' : 'Sign in'),
                    ),
                    if (_busy) ...[
                      const SizedBox(height: 10),
                      const LinearProgressIndicator(),
                      const SizedBox(height: 6),
                      Text(
                        'Contacting the server… the first sign-in can take up to a minute '
                        'while it wakes up. Please keep waiting.',
                        style: theme.textTheme.bodySmall
                            ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
                        textAlign: TextAlign.center,
                      ),
                    ],
                  ],
                ),
              ),
            ),
          ),
          Positioned(
            left: 0,
            right: 0,
            bottom: 16,
            child: Text(
              branding?.motto.isNotEmpty == true
                  ? '${branding?.name} · ${branding?.motto}'
                  : 'Powered by LIGHT TECHNOLOGIES',
              textAlign: TextAlign.center,
              style: theme.textTheme.labelSmall
                  ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
            ),
          ),
        ],
      ),
    );
  }
}

/// A "Secure connection failed" almost always means the phone's clock is wrong
/// (phones with no SIM never network-sync their time). Surface the exact fix.
class _ClockFixHint extends StatelessWidget {
  const _ClockFixHint({required this.errorText});

  final String errorText;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final hint = errorText.toLowerCase();
    if (!hint.contains('date') && !hint.contains('time') &&
        !hint.contains('secure connection')) {
      return const SizedBox.shrink();
    }
    return Padding(
      padding: const EdgeInsets.only(top: 6),
      child: Container(
        padding: const EdgeInsets.all(10),
        decoration: BoxDecoration(
          color: theme.colorScheme.errorContainer,
          borderRadius: BorderRadius.circular(4),
        ),
        child: Text(
          'Fix: Settings → System → Date & time → turn ON "Set time '
          'automatically" (or set today’s correct date), then tap Sign in again. '
          'A wrong clock blocks the secure connection.',
          style: theme.textTheme.bodySmall!.copyWith(
            color: theme.colorScheme.onErrorContainer,
          ),
        ),
      ),
    );
  }
}